xarray/core/dataset.py

```diff
@@ -2962,17 +2962,38 @@ def _validate_interp_indexer(x, new_x):
                 )
             return x, new_x
 
+        validated_indexers = {
+            k: _validate_interp_indexer(maybe_variable(obj, k), v)
+            for k, v in indexers.items()
+        }
+
+        # optimization: subset to coordinate range of the target index
+        if method in ["linear", "nearest"]:
+            for k, v in validated_indexers.items():
+                obj, newidx = missing._localize(obj, {k: v})
+                validated_indexers[k] = newidx[k]
```

Comment:
```suggestion
        validated_indexers = dict()
        method_is_localizeable = method in ["linear", "nearest"]
        for k, v in indexers.items():
            validated_indexer = _validate_interp_indexer(maybe_variable(obj, k), v)

            # optimization: subset to coordinate range of the target index:
            if method_is_localizeable:
                obj, newidx = missing._localize(obj, {k: validated_indexer})
                validated_indexer = newidx[k]
            validated_indexers[k] = validated_indexer
```
Could do something like this to avoid the mutating loops.